src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php

```diff
@@ -0,0 +1,107 @@
+<?php
+declare(strict_types=1);
+
+namespace Roave\BetterReflection\SourceLocator\Type;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitorAbstract;
+use PhpParser\ParserFactory;
+use ReflectionClass as CoreReflectionClass;
+use Roave\BetterReflection\Identifier\Identifier;
+use Roave\BetterReflection\Identifier\IdentifierType;
+use Roave\BetterReflection\Reflection\Reflection;
+use Roave\BetterReflection\Reflection\ReflectionClass;
+use Roave\BetterReflection\Reflector\Reflector;
+use Roave\BetterReflection\SourceLocator\Ast\Strategy\NodeToReflection;
+use Roave\BetterReflection\SourceLocator\Exception\TwoAnonymousClassesOnSameLine;
+use Roave\BetterReflection\SourceLocator\Located\LocatedSource;
+use Roave\BetterReflection\Util\FileHelper;
+
+final class AnonymousClassObjectSourceLocator implements SourceLocator
+{
+    /**
+     * @var CoreReflectionClass
+     */
+    private $coreClassReflection;
+
+    /**
+     * @param object $anonymousClassObject
+     */
+    public function __construct($anonymousClassObject)
+    {
+        $this->coreClassReflection = new CoreReflectionClass($anonymousClassObject);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @throws \Roave\BetterReflection\SourceLocator\Ast\Exception\ParseToAstFailure
+     */
+    public function locateIdentifier(Reflector $reflector, Identifier $identifier) : ?Reflection
+    {
+        return $this->getReflectionClass($reflector, $identifier->getType());
+    }
+
+    /**
+     * {@inheritDoc}
+     * @throws \Roave\BetterReflection\SourceLocator\Ast\Exception\ParseToAstFailure
+     */
+    public function locateIdentifiersByType(Reflector $reflector, IdentifierType $identifierType) : array
+    {
+        return array_filter([$this->getReflectionClass($reflector, $identifierType)]);
+    }
+
+    private function getReflectionClass(Reflector $reflector, IdentifierType $identifierType) : ?ReflectionClass
+    {
+        if (!$identifierType->isClass()) {
+            return null;
+        }
+
+        $fileName = FileHelper::normalizeWindowsPath($this->coreClassReflection->getFileName());
+
+        $ast = (new ParserFactory())->create(ParserFactory::PREFER_PHP7)->parse(file_get_contents($fileName));
```

Comment:
I added commit with `FileChecker`. I've placed it to `SourceLocator` namespace because the `InvalidFileLocation` is also there. However it may be better to place the class to `Util` namespace or simple add `checkFile` method to already existing `FileHelper`.